FRAuth/FRAuth/DeviceBinding/DeviceBindingAuthenticators.swift

```diff
@@ -114,180 +125,222 @@ extension DeviceAuthenticator {
-        let keyBuilderQuery = keyAware.keyBuilderQuery()
-        return try keyAware.createKeyPair(builderQuery: keyBuilderQuery)
+    open func generateKeys() throws -> KeyPair {
+        guard let cryptoKey = cryptoKey else {
+            throw DeviceBindingStatus.unsupported(errorMessage: "Cannot generate keys, missing cryptoKey")
+        }
+        
+        let keyBuilderQuery = cryptoKey.keyBuilderQuery()
+        return try cryptoKey.createKeyPair(builderQuery: keyBuilderQuery)
     }
     
     
     /// Check if authentication is supported
-    func isSupported() -> Bool {
+    open func isSupported() -> Bool {
         return true
     }
     
     
     /// Access Control for the authetication type
-    func accessControl() -> SecAccessControl? {
+    open func accessControl() -> SecAccessControl? {
         return nil
     }
+    
+    
+    open func type() -> DeviceBindingAuthenticationType {
+        return .none
+    }
+    
+    open func setKey(cryptoKey: CryptoKey) {
+        self.cryptoKey = cryptoKey
+    }
 }
 
 
-///Public and private keypair struct
-public struct KeyPair {
-    /// The Private key
-    var privateKey: SecKey
-    /// The Private key
-    var publicKey: SecKey
-    /// Alias for the key
-    var keyAlias: String
-}
-
-
-/// AuthenticatorFactory to create the authentication type.
-internal struct AuthenticatorFactory {
-    
-    /// Static method to create and return the correct type of authenticator
-    static func getAuthenticator(userId: String,
-                                 authentication: DeviceBindingAuthenticationType,
-                                 title: String,
-                                 subtitle: String,
-                                 description: String,
-                                 keyAware: KeyAware?) -> DeviceAuthenticator {
-        let newKeyAware = keyAware ?? KeyAware(userId: userId)
-        switch authentication {
+/// Convert authentication type string received from server to authentication type enum
+public enum DeviceBindingAuthenticationType: String, Codable {
+    case biometricOnly = "BIOMETRIC_ONLY"
+    case biometricAllowFallback = "BIOMETRIC_ALLOW_FALLBACK"
+    case applicationPin = "APPLICATION_PIN"
+    case none = "NONE"
+    
+    /// get the right type of DeviceAuthenticator
+    func getAuthType() -> DeviceAuthenticator {
+        switch self {
         case .biometricOnly:
-            return BiometricOnly(description: description, keyAware: newKeyAware)
+            return BiometricOnly()
         case .biometricAllowFallback:
-            return BiometricAndDeviceCredential(description: description, keyAware: newKeyAware)
+            return BiometricAndDeviceCredential()
+        case .applicationPin:
+            return ApplicationPinDeviceAuthenticator()
         case .none:
-            return None(keyAware: newKeyAware)
-            
+            return None()
         }
     }
 }
+
+
+public struct Prompt {
+    var title: String
```

Comment:
I would consider having those as optional strings to avoid passing empty string values when not needed

Reply:
@george-bafaloukas-forgerock I thought about it, but I am using the empty strings only in the tests. In the callbacks where it's used we have all 3 property values.